libfreerdp/core/gateway/rdg.c

```diff
@@ -147,6 +148,70 @@ typedef struct rdg_packet_header
 
 #pragma pack(pop)
 
+typedef struct
+{
+	UINT32 code;
+	const char* name;
+} t_err_mapping;
+
+static const t_err_mapping fields_present[] =
+{
+	{HTTP_CHANNEL_RESPONSE_FIELD_CHANNELID,   "HTTP_CHANNEL_RESPONSE_FIELD_CHANNELID"},
+	{HTTP_CHANNEL_RESPONSE_OPTIONAL, "HTTP_CHANNEL_RESPONSE_OPTIONAL"},
+	{HTTP_CHANNEL_RESPONSE_FIELD_UDPPORT,     "HTTP_CHANNEL_RESPONSE_FIELD_UDPPORT"}
+};
+
+static const t_err_mapping extended_auth[] =
+{
+	{HTTP_EXTENDED_AUTH_NONE, "HTTP_EXTENDED_AUTH_NONE"},
+	{HTTP_EXTENDED_AUTH_SC, "HTTP_EXTENDED_AUTH_SC"},
+	{HTTP_EXTENDED_AUTH_PAA, "HTTP_EXTENDED_AUTH_PAA"},
+	{HTTP_EXTENDED_AUTH_SSPI_NTLM, "HTTP_EXTENDED_AUTH_SSPI_NTLM"}
+};
+
+static const char* fields_present_to_string(UINT16 fieldsPresent)
+{
+	size_t x = 0;
+	static char buffer[1024] = { 0 };
+	char fields[12];
+
+	for (x = 0; x < ARRAYSIZE(fields_present); x++)
+	{
+		if (strnlen(buffer, ARRAYSIZE(buffer)) > 0)
+			strcat(buffer, "|");
+
+		if ((fields_present[x].code & fieldsPresent) != 0)
+			strcat(buffer, fields_present[x].name);
+	}
+
+	sprintf_s(fields, ARRAYSIZE(fields), " [%04"PRIx16"]", fieldsPresent);
+	strcat(buffer, fields);
+	return buffer;
+}
+
+static const char* extended_auth_to_string(UINT16 auth)
+{
+	size_t x = 0;
+	static char buffer[1024] = { 0 };
+	char fields[12];
+
+	if (auth == HTTP_EXTENDED_AUTH_NONE)
+		return "HTTP_EXTENDED_AUTH_NONE [0x0000]";
+
+	for (x = 0; x < ARRAYSIZE(extended_auth); x++)
+	{
+		if (strnlen(buffer, ARRAYSIZE(buffer)) > 0)
+			strcat(buffer, "|");
```

Comment:
Potential buffer overflow (writes at least 2 bytes, strnlen checks for > 0)

Reply:
No, the buffer is large enough to hold all values.